### ISSUE TITLE
[Bug] Constraints not enforced where `workspace_field ` is absent

### ISSUE THREAD

**Issue by theseyi** [yarnpkg/berry#2419]
- [ ] I'd be willing to implement a fix

**Describe the bug**
Creating a constraint using the query predicate to enforce the fact for workspace manifests where a field does _not_ exist.
Running `yarn constraints` does not enforce the constraint.

**To Reproduce**
Seem minimal reproduction here https://github.com/theseyi/constraints

- init a yarn workspace with one workspace package `packages/package-a` and the workspace root
- import the constraint plugin `yarn plugin import constraints`
- touch a constraints.pro file and add in the following constraint
- attempt to enforce the constraint with `yarn constraints`
```pro repro
% enforce workspace packages without a workspace field to have a scripts.test field
gen_enforced_field(WorkspaceCwd, 'scripts.test', 'someValue') :-
  workspace_field(WorkspaceCwd, 'workspaces', null).
```
<details>
  <summary>Reproduction</summary>

```json repro
// in workspace root
{
  "name": "constraints",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "workspaces": [
    "packages/*"
  ]
}
```

```json repro
// in packages/package-a
{
  "name": "package-a",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT"
}
```
```pro repro
% enforce workspace packages without a workspace field to have a scripts.test field
gen_enforced_field(WorkspaceCwd, 'scripts.test', 'someValue') :-
  workspace_field(WorkspaceCwd, 'workspaces', null).
```
</details>

 *Expected*:
```sh
➤ YN0037: package-a must have a field scripts.test set to someValue, but doesn't
➤ YN0000: Failed with errors in 0s 18ms
```
**IMPORTANT**: We **strongly** prefer reproductions that use Sherlock. Please check our documentation
for more information: https://yarnpkg.com/advanced/sherlock

_Unable to use Sherlock here since I'm not familiar with running it over prolog_

**Environment if relevant (please complete the following information):**

 - OS: `macos`
 - Node version `v12.20.1`
 - Yarn version `2.4.0`

**Comment by merceyz** [comment]
You can update your constraint to check that it has any value and then negate it to get the result you're after

```prolog
gen_enforced_field(WorkspaceCwd, 'scripts.test', 'someValue') :-
  \+ workspace_field(WorkspaceCwd, 'workspaces', _).
```

**Comment by theseyi** [comment]
Awesome thank you @merceyz 🙏. Although, shouldn't the `null` value indicate that the field is absent and cause the constraint to be enforced?

**Comment by theseyi** [comment]
Closing since there is a workaround, but does the original bug still seem valid? @merceyz

**Comment by merceyz** [comment]
> Although, shouldn't the null value indicate that the field is absent and cause the constraint to be enforced?

No, not in this case. `null` is only special in the `gen_*` predicates so we should document the negation part, would you like to submit a PR?

**Comment by theseyi** [comment]
Sure, will get to it today, thanks for following up
